### PR TITLE
DEP: Deprecate non-tuple nd-indices

### PR DESCRIPTION
**Pull request by eric-wieser**
A reminder of the purpose of this - currently we allow both `arr[[None, 0]]` and `arr[(None, 0)]` to mean the same thing, yet `arr[[0, 0]]` and `arr[(0, 0)]` mean different things. This makes it super hard to make a compliant subclass, or duck array.

By deprecating this feature, we force downstream library code to stop using it, which in turn makes that library code more compatible with subclasses and duck types.

---

This reapplies @seberg's #4434, with the following changes:

* `tuple` subclasses are not deprecated, because supporting `namedtuple`s seems possibly useful
* All warnings are `FutureWarnings`, since determining how `np.array(...)` will treat a sequence is expensive - it's easier to just say _"this will error if `np.array(seq)` would"_

As of #10618, none of our code relies on this "feature"

### CODE REVIEW

File: numpy/fft/fftpack.py [hunk 1]
```diff
@@ -69,13 +69,13 @@ def _raw_fft(a, n=None, axis=-1, init_function=fftpack.cffti,
         if s[axis] > n:
             index = [slice(None)]*len(s)
```
Comment (mhvk) [inline numpy/fft/fftpack.py]: Maybe nicer to just fix the creation of the index:
```
index = (slice(None),) * axis + (slice(0, n),)
```
then, no need to use `tuple` below.

File: doc/release/1.14.0-notes.rst [hunk 2]
```diff
@@ -21,6 +21,12 @@ New functions
 Deprecations
 ============
 
+* Multidimensional indexing with anything but a base class tuple is
+  deprecated. This means that code such as ``arr[[slice(None)]]`` has to
```
Comment (mhvk) [inline doc/release/1.14.0-notes.rst]: How about giving an example with a slightly more useful case, e.g., `[slice(None), slice(0, 10)]` -> `tuple(slice(None), slice(0, 10))` (or even omit the tuple in that case).

File: numpy/core/tests/test_memmap.py [hunk 3]
```diff
@@ -126,7 +126,7 @@ def test_arithmetic_drops_references(self):
     def test_indexing_drops_references(self):
         fp = memmap(self.tmpfp, dtype=self.dtype, mode='w+',
                     shape=self.shape)
-        tmp = fp[[(1, 2), (2, 3)]]
+        tmp = fp[((1, 2), (2, 3))]
```
Comment (mhvk) [inline numpy/core/tests/test_memmap.py]: Here, the parenthesis are not necessary, correct?

File: numpy/fft/fftpack.py [hunk 4]
```diff
@@ -69,13 +69,13 @@ def _raw_fft(a, n=None, axis=-1, init_function=fftpack.cffti,
...
             index[axis] = slice(0, n)
-            a = a[index]
+            a = a[tuple(index)]
         else:
             index = [slice(None)]*len(s)
             index[axis] = slice(0, s[axis])
```
Comment (mhvk) [inline numpy/fft/fftpack.py]: And same here.
```
index = (slice(None),) * axis + (slice(0, s[axis]),)
```

File: numpy/lib/function_base.py [hunk 5]
```diff
@@ -1745,7 +1745,7 @@ def gradient(f, *varargs, **kwargs):
         slice4[axis] = slice(2, None)
```
Comment (mhvk) [inline numpy/lib/function_base.py]: If up to me, I'd just create a base prefix `(slice(None),) * axis` here and add the 1,2,3,4 as need below - which I think would make for clearer code. But that is getting well beyond the immediate purpose of this PR, so feel free to ignore.

File: numpy/lib/function_base.py [hunk 6]
```diff
@@ -2159,6 +2159,7 @@ def unwrap(p, discont=pi, axis=-1):
     dd = diff(p, axis=axis)
     slice1 = [slice(None, None)]*nd     # full slices
     slice1[axis] = slice(1, None)
+    slice1 = tuple(slice1)
```
Comment (mhvk) [inline numpy/lib/function_base.py]: Could just be
```
slice1 = (slice(None),) * axis + (slice(1, None),)
```
I'm starting to wonder if we shouldn't have a small helper function `slice_along_axis`...

File: numpy/lib/function_base.py [hunk 7]
```diff
@@ -4141,6 +4142,7 @@ def _median(a, axis=None, out=None, overwrite_input=False):
         indexer[axis] = slice(index, index+1)
     else:
         indexer[axis] = slice(index-1, index+1)
+    indexer = tuple(indexer)
```
Comment (mhvk) [inline numpy/lib/function_base.py]: Again no need for the list:
```
indexer = (slice(None),) * axis
if part.shape[axis] % 2 == 1:
    indexer += (slice(index, index+1),)
else:
    indexer += (slice(index-1, index+1),)
```

File: numpy/lib/function_base.py [hunk 8]
```diff
@@ -4497,12 +4499,12 @@ def trapz(y, x=None, dx=1.0, axis=-1):
     slice1[axis] = slice(1, None)
     slice2[axis] = slice(None, -1)
```
Comment (mhvk) [inline numpy/lib/function_base.py]: Same thing again...

File: numpy/lib/function_base.py [hunk 9]
```diff
@@ -4791,15 +4793,15 @@ def delete(arr, obj, axis=None):
             pass
         else:
             slobj[axis] = slice(None, start)
-            new[slobj] = arr[slobj]
+            new[tuple(slobj)] = arr[tuple(slobj)]
```
Comment (mhvk) [inline numpy/lib/function_base.py]: Probably going out of PR again, but one could define `slobj` as a prefix tuple and just add the relevant slice here.

File: numpy/ma/core.py [hunk 10]
```diff
@@ -5483,6 +5483,7 @@ def sort(self, axis=-1, kind='quicksort', order=None,
         else:
             idx = list(np.ix_(*[np.arange(x) for x in self.shape]))
             idx[axis] = sidx
+            idx = tuple(idx)
```
Comment (mhvk) [inline numpy/ma/core.py]: Maybe
```
idx = np.ix_(*[(sidx if i == axis else np.arange(self.shape[i])) for i in range(self.ndim)])
```
well, maybe not really better at all.

File: numpy/ma/extras.py [hunk 11]
```diff
@@ -723,6 +723,7 @@ def _median(a, axis=None, out=None, overwrite_input=False):
         # as median (which is mean of empty slice = nan)
         indexer = [slice(None)] * asorted.ndim
         indexer[axis] = slice(0, 0)
+        indexer = tuple(indexer)
```
Comment (mhvk) [inline numpy/ma/extras.py]:
```
indexer = (slice(None),) * axis + (slice(0, 0),)
```

File: numpy/ma/extras.py [hunk 12]
```diff
@@ -784,6 +785,7 @@ def replace_masked(s):
     if np.issubdtype(asorted.dtype, np.inexact):
         # avoid inf / x = masked
         s = np.ma.sum([low, high], axis=0, out=out)
+        print(repr(s.data))
```
Comment (mhvk) [inline numpy/ma/extras.py]: Probably not intended!

File: numpy/ma/extras.py [hunk 13]
```diff
@@ -1657,7 +1659,7 @@ def flatnotmasked_contiguous(a):
         if not k:
             result.append(slice(i, i + n))
         i += n
-    return result or None
+    return tuple(result) or None
```
Comment (mhvk) [inline numpy/ma/extras.py]: Also update the docstring! (It states a list is returned)

File: numpy/ma/extras.py [hunk 14]
```diff
@@ -1715,8 +1717,8 @@ def notmasked_contiguous(a, axis=None):
     #
     for i in range(a.shape[other]):
         idx[other] = i
-        result.append(flatnotmasked_contiguous(a[idx]) or None)
-    return result
+        result.append(flatnotmasked_contiguous(a[tuple(idx)]) or None)
+    return tuple(result)
```
Comment (mhvk) [inline numpy/ma/extras.py]: Again do update docstring.

**Review by mhvk — COMMENTED** [review]
Most comments just attempt at making a tuple to start with. The many occurrences in our own code do make me wonder slightly whether this is in fact a good idea. I fear the construction of lists consisting of lots of `slice(None)` plus a real `slice` is going to be common also outside of `numpy`...

File: numpy/ma/extras.py (hunk 12 above)
Comment (eric-wieser) [reply numpy/ma/extras.py]: Nice catch! Must have been in my working copy or stash or something

**Comment by eric-wieser** [comment]
Just some notes from a performance perspective:

* until #8278, passing lists was probably slower anyway, because you ended up doing a bunch of heuristics and then calling `PyTuple_new` on the list on the C side.
* coercing a list to a tuple seems faster than building a tuple from scratch each time:
```
>>> a = np.zeros((2,)*10)

>>> sln = (slice(None),)
>>> %timeit a[sln*3 + (1,) + sln*6]
933 ns ± 20 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

>>> sl = list(sln * 10)
>>> %timeit sl[3] = 1; a[sl]
748 ns ± 27.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
>>> %timeit sl[3] = 1; a[tuple(sl)]
858 ns ± 14.7 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

**Comment by seberg** [comment]
Fun, it used to be quite a bit slower, I guess python optimized the
keyword parsing in `tuple` or something.

**Comment by eric-wieser** [comment]
Weirdly, writing it out explicitly is slowest of all:

```
In [42]: %timeit a[:,:,:,1,:,:,:,:,:,:]
985 ns ± 10.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

**Comment by seberg** [comment]
Dunno, might be unfair since python has to create the slice objects inside the timeit, etc.

**Comment by mhvk** [comment]
The performance aspect probably doesn't matter so much; my comments were more about trying to make the code cleaner, with less conversions.

But I must admit that I've wondered why direct list comprehension is so much faster than doing it inside a function:
```
In [1]: %timeit [_i for _i in range(100)]
100000 loops, best of 3: 5.75 µs per loop

In [2]: %timeit list(_i for _i in range(100))
100000 loops, best of 3: 9.94 µs per loop

In [3]: %timeit tuple(_i for _i in range(100))
100000 loops, best of 3: 10.1 µs per loop
```

**Comment by charris** [comment]
Sounds like this might need more discussion.

**Comment by charris** [comment]
Pushing off to 1.15.

File: numpy/core/tests/test_memmap.py (hunk 3 above)
Comment (eric-wieser) [reply numpy/core/tests/test_memmap.py]: Fixed

File: numpy/ma/extras.py (hunk 12 above)
Comment (eric-wieser) [reply numpy/ma/extras.py]: Fixed

File: numpy/ma/extras.py (hunk 13 above)
Comment (eric-wieser) [reply numpy/ma/extras.py]: Fixed

File: numpy/ma/extras.py (hunk 14 above)
Comment (eric-wieser) [reply numpy/ma/extras.py]: Fixed

**Comment by eric-wieser** [comment]
@mhvk: I'd prefer to leave the different tuple generation strategy to a different PR, and keep this change as small as possible.

File: numpy/core/src/multiarray/mapping.c [hunk 15]
```diff
@@ -313,6 +313,15 @@ unpack_indices(PyObject *index, PyObject **result, npy_intp result_n)
...
                     || tmp_obj == Py_Ellipsis
                     || tmp_obj == Py_None) {
+                if (DEPRECATE_FUTUREWARNING(
+                        "Using a non-tuple sequence for multidimensional "
+                        "indexing is deprecated; use `arr[tuple(seq)]` "
+                        "instead of `arr[seq]`. In the future this will be "
+                        "interpreted as an array index, `arr[np.array(seq)]`, "
+                        "which will result either in an error or a different "
+                        "result.") < 0) {
+                    return -1;
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/mapping.c]: Think I might have missed some cleanup here.

**Comment by eric-wieser** [comment]
Base switch above to recompute the diff. Shame there's no button to do that.

**Comment by eric-wieser** [comment]
This maybe needs a doc change to https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html#advanced-indexing

**Comment by charris** [comment]
@eric-wieser Want to make that change or can I put this in.

**Comment by eric-wieser** [comment]
I think this can go in - the docs aren't wrong, and already mention it's there for compatibility. We can always improve the docs later.

**Comment by eric-wieser** [comment]
Ok, I improved the docs, and updated the commit message + PR title for a little more rationale.

**Comment by charris** [comment]
Thanks Eric.

**Comment by rgommers** [comment]
> Most comments just attempt at making a tuple to start with. The many occurrences in our own code do make me wonder slightly whether this is in fact a good idea.

There's a long list of occurences in scipy as well: https://github.com/scipy/scipy/pull/8879#issuecomment-392644241. I think that means that this deprecation is either not the best idea, or can at least not be turned into an exception without a major version bump.

**Comment by eric-wieser** [comment]
>  or can at least not be turned into an exception without a major version bump.

Is this possible? We could switch this to `PendingDeprecationWarning` if it helped.

Interesting that one of those cases is internal to numpy - a gap in our tests, it seems

**Comment by rgommers** [comment]
> Is this possible?

not sure what you mean with this? Is a major version bump possible? 

> We could switch this to `PendingDeprecationWarning` if it helped.

probably would help; things will get noisy quickly otherwise it seems

**Comment by eric-wieser** [comment]
What I mean is, how is it possible for us as the warning-producer to decide whether a warning gets turned into an exception? Isn't that down to the user, since by default no warnings are turned into exceptions?

**Comment by rgommers** [comment]
Ah no, I wasn't talking about the behavior of `warnings`, but that we cannot *remove* the now deprecated feature from numpy without a major version bump, because this seems like a quite common code construct.

**Comment by eric-wieser** [comment]
Oh, right. I completely agree - actually changing this behavior is very risky - but emitting the warnings now makes things easier on subclassers / duck-typers.

**Comment by eric-wieser** [comment]
So, right now the follow emits an unhelpful warning:
```
>>> m = np.matrix([1, 2])
>>> m[[slice(None)]]
site-packages\numpy\matrixlib\defmatrix.py:179: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  out = N.ndarray.__getitem__(self, index)
matrix([[1, 2]])
```

Is there any way to make this point at the offending `m[[slice(None)]]` line, rather than the pass-through `matrix.__getitem__`?

**Comment by ilayn** [comment]
`DEPRECATE_FUTUREWARNING` needs to be modified such that the `stacklevel` becomes the last argument in `PyErr_WarnEx` signature. That would allow to change the location of the emitting on the stack.

https://github.com/numpy/numpy/blob/ae940f950c2c62ae77561c913319ff25a46652e9/numpy/core/include/numpy/ndarrayobject.h#L269-L270

**Comment by seberg** [comment]
Yeah, but it is basically wild guessing what the correct stack level is here (you would have to check whether you have a subclass that implements `__getitem__`, and then hope that we came from there?). I guess it could be useful anyway if it does not get too complex....

Maybe silly, but you could add a note to the message hinting that conversion to error gives a clear traceback?

**Comment by shoyer** [comment]
Is there a plan for finishing this deprecation cycle? Would 1.17 be too soon to change this?

**Comment by eric-wieser** [comment]
My gut feel is to leave this till 1.18 - this was quite an intrusive deprecation, and the PR references above suggest that downstream packages are still running into it every few months. Perhaps the deprecation NEP has a clear stance here.

Is there some reason you're hoping to get rid of this sooner? My main goal with this PR was to stop users writing code in this style, so I could stop trying to replicate it in my own `__getitem__`, without actually caring if the change ever happens

**Comment by shoyer** [comment]
Right now, `numpy_array[duck_array]` can fail in some but not all circumstances to be interpreted as `numpy_array[np.asarray(duck_array)]`. This resulted in an issue for JAX: https://github.com/google/jax/issues/620

But I agree, it seems that we should wait a bit longer before changing this.

**Comment by eric-wieser** [comment]
Do the `jax` objects implement `__array_ufunc__`, or some other marker we could use look for meaning "do not use the deprecated behavior"?

**Comment by shoyer** [comment]
They implement `__array__` currently, and will soon implement both `__array_ufunc__` and `__array_function__` (https://github.com/google/jax/pull/611).

**Comment by eric-wieser** [comment]
I think the presence of `__array__` is probably a good indication that we should not treat them as a tuple. Perhaps add a check for that attribute to shortcut this path, and if the type is not `list`?

**Comment by rgommers** [comment]
> Perhaps the deprecation NEP has a clear stance here.

It won't have an opinion on any specific cases. That was a clear message from review: just express the principles not current cases. I'm planning to get back to that NEP soon by the way.

I agree with waiting till 1.18

**Comment by eric-wieser** [comment]
We might be able to finish the deprecation now for everything except `list`s.

We could also consider bumping this to an `np.VisibleDeprecationWarning` to try and flush out remaining downstream projects.

**Comment by rgommers** [comment]
> We could also consider bumping this to an np.VisibleDeprecationWarning to try and flush out remaining downstream projects.

I like that idea, if we can do that for 1.17 that will probably flush out many more issues for end users.

**Pull request merged**